pkg/template/jsonpath.go

```diff
@@ -0,0 +1,158 @@
+// If the value if an array or map, it returns a JSON representation
+// of the value (as opposed to the internal go representation of the value)
+// Otherwise, the text value is from the `evalToText` function, originally from
+// k8s.io/client-go/util/jsonpath/jsonpath.go
+func textValue(v reflect.Value) ([]byte, error) {
+	t := reflect.TypeOf(v.Interface())
+	// special case for null values in JSON; evalToText() returns <nil> here
+	if t == nil {
+		return []byte("null"), nil
+	}
+
+	switch t.Kind() {
+	// evalToText() returns <map> ....; return JSON string instead.
+	case reflect.Map, reflect.Slice:
+		return json.Marshal(v.Interface())
+	default:
+		return evalToText(v)
+	}
+}
+
+// evalToText translates reflect value to corresponding text
+// This is a unmodified copy of j.evalToText from k8s.io/client-go/util/jsonpath/jsonpath.go
+func evalToText(v reflect.Value) ([]byte, error) {
+	iface, ok := template.PrintableValue(v)
+	if !ok {
+		return nil, fmt.Errorf("can't print type %s", v.Type())
+	}
+	var buffer bytes.Buffer
+	fmt.Fprint(&buffer, iface)
+	return buffer.Bytes(), nil
+}
+
+// TektonJSONPathExpression returns a valid JSONPath expression. It accepts
+// a "RelaxedJSONPath" expression that is wrapped in the Tekton variable
+// interpolation syntax i.e. $(). RelaxedJSONPath expressions can optionally
+// omit the leading curly braces '{}' and '.'
+func TektonJSONPathExpression(expr string) (string, error) {
+	if !isTektonExpr(expr) {
+		return "", errors.New("Expression not wrapped in $()")
+	}
+	return relaxedJSONPathExpression(unwrapTektonExpr(expr))
+}
+
+// RelaxedJSONPathExpression attempts to be flexible with JSONPath expressions, it accepts:
+//   * metadata.name (no leading '.' or curly braces '{...}'
+//   * {metadata.name} (no leading '.')
+//   * .metadata.name (no curly braces '{...}')
+//   * {.metadata.name} (complete expression)
+// And transforms them all into a valid jsonpath expression:
+//   {.metadata.name}
+// This function has been copied as-is from
+// https://github.com/kubernetes/kubectl/blob/c273777957bd657233cf867892fb061a6498dab8/pkg/cmd/get/customcolumn.go#L47
+func relaxedJSONPathExpression(pathExpression string) (string, error) {
+	if len(pathExpression) == 0 {
+		return pathExpression, nil
+	}
+	submatches := jsonRegexp.FindStringSubmatch(pathExpression)
+	if submatches == nil {
+		return "", fmt.Errorf("unexpected path string, expected a 'name1.name2' or '.name1.name2' or '{name1.name2}' or '{.name1.name2}'")
+	}
+	if len(submatches) != 3 {
+		return "", fmt.Errorf("unexpected submatch list: %v", submatches)
+	}
+	var fieldSpec string
+	if len(submatches[1]) != 0 {
+		fieldSpec = submatches[1]
+	} else {
+		fieldSpec = submatches[2]
+	}
+	return fmt.Sprintf("{.%s}", fieldSpec), nil
+}
+
+// unwrapTektonExpr takes an expression of the form $(abc) and returns abc.
+func unwrapTektonExpr(expr string) string {
+	return tektonVar.ReplaceAllString(expr, "$1")
```

Comment:
im not a huge fan of functions that only have one line in them, i feel like they just add a layer of indirection without adding too much (e.g. if i want to know what `unwrapTektonExpr` does i have to come over here and look at this vs looking direclty at `tektonVar.ReplaceAllString`), but maybe there is a benefit i'm missing?

Reply:
removed